### ISSUE TITLE
properly stacking BufferingWrapper and RetryingWrapper

### ISSUE THREAD

**Issue by WhatWapp** [NLog/NLog#1869]
**Type** (choose one):
- Question

**NLog version**: (e.g.4.3.11)

**Platform**: .Net 3.5 / Unity3d

Hi, is there a way to stack buffering and retrying wrappers (over a custom target) such that:
1. each log message gets buffered
2. when buffering triggers (e.g. N entries or timeout) it gets sent over network
3. if the request fails, the entire batch gets sent again
4. retrying doesn't block the main thread (i have to start network requests in the main thread because of Unity's WWW)

or it's better to write the retry logic directly inside my target?
ps. I'm overriding:
```c#
protected override void Write (AsyncLogEventInfo logEvent){
	Write (new[]{ logEvent });
}
protected override void Write (AsyncLogEventInfo[] logEvents){
// send string.Join("\n", logEvents.Select (ale => Layout.Render (ale.LogEvent)).ToArray ()) to remote
}
```

**Comment by snakefoot** [comment]
RetryingTargetWrapper doesn't support retrying the entire LogEvents-array. It would also require that it could see that the wrapped-target was able to handle an entire LogEvents-arrays.

Since you are writing a custom target, then I would probably implement retry logic yourself. Inside the method taking the LogEvents-array.

**Comment by WhatWapp** [comment]
thanks. it seems I have to add it inside also because of the main thread problem (RetryingTargetWrapper just thread.sleeps when I put a delay)

btw I see that the default implementation of Write(array) just loops through the events and calls write on each one. it's not possible to add array support to the retrying wrapper by implementing the logic in the array overload and letting the wrapped target decide if it supports it? if not, what would be the recommended way to send buffered logs over a remote target and have it retry on error? (e.g. buffer->retry or retry->buffer?)

**Comment by snakefoot** [comment]
I guess in the ideal world then it would be:

- BufferingTargetWrapper -> RetryingTargetWrapper -> FinalTarget

Sadly enough the BufferingTargetWrapper takes a target-lock when timer-thread fires. This will block for any new LogEvents while waiting inside RetryingTargetWrapper  (This is fixed with #1853)

Also sadly enough there is no property on RetryingTargetWrapper called AllowBatchRetry, that allows it to forward the entire LogEvents-array-batch to the WrappedTarget (instead of iterating the array).

So right now you have to do the following:

- AsyncTargetWrapper -> BufferingTargetWrapper -> FinalTarget with custom batch retry

**Comment by WhatWapp** [comment]
ok thanks.

**Comment by 304NotModified** [comment]
> Sadly enough the BufferingTargetWrapper takes a target-lock when timer-thread fires. This will block for any new LogEvents while doing thread-sleep inside RetryingTargetWrapper (This is fixed with #1853)

Thanks!


> Also sadly enough there is no property on RetryingTargetWrapper called AllowFullBatchRetry, that allows it to forward the entire LogEvents-array-batch to the WrappedTarget (instead of iterating the array), and retrying the entire LogEvents-array-batch on error.

Guess that one is not to difficult to implement? Could you make a new issue for this? Thanks in advan
